### PR TITLE
Restructure Spline class to be compatible with different type of Eigen Vectors

### PR DESCRIPTION
**Pull request by GiulioRomualdi**
This pull request brings about a restructuring of the spline class, ensuring compatibility with various types of eigen vectors. The specific changes include:

1. The relocation of the spline class to the math component.
2. To maintain backward compatibility in both C++ and Python, the previous class has not been removed but instead deprecated.
4. Elimination of code redundancy in both CubicSpline and QuinticSpline implementations.
5. Implement the possibility to efficiently query the points of a spline given an ordered time vector

### CODE REVIEW

File: bindings/python/Math/src/Spline.cpp [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+#include <BipedalLocomotion/bindings/System/Advanceable.h>
+
+namespace BipedalLocomotion
+{
+namespace bindings
+{
+namespace Math
+{
+
+template <typename T> void CerateTrajectoryPoint(pybind11::module& module, const std::string& name)
```
Comment (S-Dafarra) [inline bindings/python/Math/src/Spline.cpp:30]: Typo ``Cerate``

File: src/Math/tests/SplineTest.cpp [hunk 2]
```diff
@@ -154,38 +189,40 @@ TEST_CASE("Cubic spline")
...
-    CubicSpline spline;
+    Eigen::Vector4d initAcceleration = 2 * coefficients[2] + 3 * 2 * coefficients[3] * initT;
+    Eigen::Vector4d finalAcceleration = 2 * coefficients[2] + 3 * 2 * coefficients[3] * finalT;
+
+    CubicSpline<Eigen::Vector4d> spline;
     REQUIRE(spline.setKnots(knots, time));
 
-    REQUIRE(spline.setInitialConditions(initVelocity));
-    REQUIRE(spline.setFinalConditions(finalVelocity));
+    REQUIRE(spline.setInitialConditions({initVelocity, initAcceleration}));
```
Comment (S-Dafarra) [inline src/Math/tests/SplineTest.cpp:222]: Is this a breaking change?

File: src/Math/include/BipedalLocomotion/Math/Spline.h [hunk 3]
```diff
@@ -0,0 +1,722 @@
...
+ * @note The spline is defined as a set of polynomials. The coefficients of the polynomials are
+ * computed using the boundary conditions. The boundary conditions are the initial and final
+ * velocity and acceleration.
+ * @tparam T type of the spline. T must be an Eigen type (i.e. Eigen::Vector3d).
+ */
+template <typename T> class Spline : public System::Source<TrajectoryPoint<T>>
+{
+
+    static_assert(is_base_of_template<Eigen::MatrixBase, T>::value,
+                  "T must be an Eigen type (i.e. Eigen::Vector3d)");
```
Comment (S-Dafarra) [inline src/Math/include/BipedalLocomotion/Math/Spline.h:47]:
```suggestion
                  "T must be an Eigen type (e.g. Eigen::Vector3d)");
```

File: src/Math/include/BipedalLocomotion/Math/Spline.h [hunk 4]
```diff
@@ -0,0 +1,722 @@
...
+    };
+
+    /**
+     * Destructor.
+     */
+    virtual ~Spline() = default;
+
+    /**
+     * Set the time step of the advance interface.
+     * @warning if the the time step is not set the user cannot use the advance features.
```
Comment (S-Dafarra) [inline src/Math/include/BipedalLocomotion/Math/Spline.h:66]:
```suggestion
     * @warning if the the time step is not set, the user cannot use the advance features.
```

File: src/Math/include/BipedalLocomotion/Math/Spline.h [hunk 5]
```diff
@@ -0,0 +1,722 @@
...
+            poly->getPosition(finalTime, position[i]);
+            poly->getVelocity(finalTime, velocity[i]);
+            poly->getAcceleration(finalTime, acceleration[i]);
+            continue;
+        }
+
+        if (poly != m_polynomials.end() && t[i] >= poly->finalPoint->time)
+        {
+            // advance the iterator
+            poly = std::find_if(poly, m_polynomials.end(), [&t, &i](const auto& p) {
```
Comment (S-Dafarra) [inline src/Math/include/BipedalLocomotion/Math/Spline.h:649]: Nitpick
```suggestion
            poly = std::find_if(++poly, m_polynomials.end(), [&t, &i](const auto& p) {
```
In theory the condition would always fail on the first iteration. I don't think there is any gain anyhow :grin:

File: src/Math/include/BipedalLocomotion/Math/CubicSpline.h [hunk 6]
```diff
@@ -0,0 +1,352 @@
...
+#include <Eigen/Sparse>
+
+#include <BipedalLocomotion/Math/Spline.h>
+
+namespace BipedalLocomotion
+{
+namespace Math
+{
+/**
+ * Cubic spline implement a 3-rd order polynomial spline in \$f\mathbb{R}^n\$f.
```
Comment (S-Dafarra) [inline src/Math/include/BipedalLocomotion/Math/CubicSpline.h:25]:
```suggestion
 * Cubic spline implements a 3-rd order polynomial spline in \$f\mathbb{R}^n\$f.
```

File: src/Math/include/BipedalLocomotion/Math/QuinticSpline.h [hunk 7]
```diff
@@ -0,0 +1,418 @@
...
+#include <Eigen/Sparse>
+
+#include <BipedalLocomotion/Math/Spline.h>
+
+namespace BipedalLocomotion
+{
+namespace Math
+{
+/**
+ * Quintic spline implement a 5-th order polynomial spline in \$f\mathbb{R}^n\$f.
```
Comment (S-Dafarra) [inline src/Math/include/BipedalLocomotion/Math/QuinticSpline.h:25]:
```suggestion
 * Quintic spline implements a 5-th order polynomial spline in \$f\mathbb{R}^n\$f.
```

File: src/Math/include/BipedalLocomotion/Math/Spline.h (hunk 5 above)
Comment (S-Dafarra) [reply src/Math/include/BipedalLocomotion/Math/Spline.h:649]: Another possibility could also be to simply remove the ``if`` above.

File: bindings/python/Math/src/Spline.cpp (hunk 1 above)
Comment (GiulioRomualdi) [reply bindings/python/Math/src/Spline.cpp]: https://github.com/ami-iit/bipedal-locomotion-framework/pull/773/commits/38a633e7e370f2dd7f715b8c02fcacc8334f3bca

File: src/Math/tests/SplineTest.cpp (hunk 2 above)
Comment (GiulioRomualdi) [reply src/Math/tests/SplineTest.cpp:222]: yes I will add it as deprecated

File: src/Math/include/BipedalLocomotion/Math/Spline.h (hunk 5 above)
Comment (GiulioRomualdi) [reply src/Math/include/BipedalLocomotion/Math/Spline.h]: I applied you suggestion in https://github.com/ami-iit/bipedal-locomotion-framework/pull/773/commits/8e37f1e43a888c7cf6fc2bed833b59d8c5d47a37

File: src/Math/tests/SplineTest.cpp (hunk 2 above)
Comment (GiulioRomualdi) [reply src/Math/tests/SplineTest.cpp:222]: Done in https://github.com/ami-iit/bipedal-locomotion-framework/pull/773/commits/8e37f1e43a888c7cf6fc2bed833b59d8c5d47a37

**Comment by GiulioRomualdi** [comment]
Hi @S-Dafarra I applied the modification you suggested

**Review by S-Dafarra — APPROVED** [review]

**Pull request merged**